File: src/tree/bintree/BinTreePrinter.java
package tree.bintree;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Reference:
 * https://stackoverflow.com/questions/4965335/how-to-print-binary-tree-diagram
 */

public class BinTreePrinter {

    public static <T> void printNode(BinTreeNode<T> root) {
        int maxLevel = maxLevel(root);

        printNodeInternal(Collections.singletonList(root), 1, maxLevel);
    }

    private static <T> void printNodeInternal(List<BinTreeNode<T>> nodes,
                                              int level,
                                              int maxLevel) {
        if (nodes.isEmpty() || isAllElementsNull(nodes)) {
            return;
        }

        int floor = maxLevel - level;
        int endgeLines = (int) Math.pow(2, (Math.max(floor - 1, 0)));
        int firstSpaces = (int) Math.pow(2, (floor)) - 1;
        int betweenSpaces = (int) Math.pow(2, (floor + 1)) - 1;

        printWhitespaces(firstSpaces);

        List<BinTreeNode<T>> newNodes = new ArrayList<>();
        for (BinTreeNode<T> binTreeNode : nodes) {
            if (binTreeNode != null) {
                System.out.print(binTreeNode.getData());
                newNodes.add(binTreeNode.getLeft());
                newNodes.add(binTreeNode.getRight());
            } else {
                newNodes.add(null);
                newNodes.add(null);
                System.out.print(" ");
            }

            printWhitespaces(betweenSpaces);
        }
        System.out.println("");

        for (int i = 1; i <= endgeLines; i++) {
            for (int j = 0; j < nodes.size(); j++) {
                printWhitespaces(firstSpaces - i);
                if (nodes.get(j) == null) {
                    printWhitespaces(endgeLines + endgeLines + i + 1);
                    continue;
                }

                if (nodes.get(j).getLeft() != null) {
                    System.out.print("/");
                } else {
                    printWhitespaces(1);
                }

                printWhitespaces(i + i - 1);

                if (nodes.get(j).getRight() != null) {
                    System.out.print("\\");
                } else {
                    printWhitespaces(1);
                }

                printWhitespaces(endgeLines + endgeLines - i);
            }
            System.out.println("");
        }
        printNodeInternal(newNodes, level + 1, maxLevel);
    }

    private static void printWhitespaces(int count) {
        for (int i = 0; i < count; i++)
            System.out.print(" ");
    }

    private static <T> int maxLevel(BinTreeNode<T> binTreeNode) {
        if (binTreeNode == null) {
            return 0;
        }

        return Math.max(maxLevel(binTreeNode.getLeft()),
                maxLevel(binTreeNode.getRight())) + 1;
    }

    private static <T> boolean isAllElementsNull(List<T> list) {
        for (Object object : list) {
            if (object != null)
                return false;
        }
        return true;
    }
}

File: src/tree/bintree/BinTreePrinter.kt
package tree.bintree


import util.max
import java.util.ArrayList


/**
 * Binary Tree Printer
 * @author MightyPork
 *
 * Reference:
 * https://stackoverflow.com/questions/4965335/how-to-prettyPrintTree-binary-tree-diagram
 */
/**
 * Node that can be printed
 */

interface PrintableNode {
	val left: PrintableNode?
	val right: PrintableNode?
	val text: String
}


/**
 * Print a tree
 *
 * @param root tree root node
 */
fun prettyPrintTree(root: PrintableNode) {
	val lines = ArrayList<List<String?>>()

	var level: MutableList<PrintableNode?> = ArrayList()
	var next: MutableList<PrintableNode?> = ArrayList()

	level.add(root)
	var nn = 1

	var widest = 0

	while (nn != 0) {
		val line = ArrayList<String?>()
		nn = 0
		for (n in level) {
			if (n == null) {
				line.add(null)
				next.add(null)
				next.add(null)
			} else {
				val aa = n.text
				line.add(aa)
				widest = max(widest, aa.length)

				next.add(n.left)
				next.add(n.right)

				if (n.left != null) {
					nn++
				}
				if (n.right != null) {
					nn++
				}
			}
		}

		if (widest % 2 == 1) {
			widest++
		}

		lines.add(line)

		val tmp = level
		level = next
		next = tmp
		next.clear()
	}

	var perpiece = lines[lines.size - 1].size * (widest + 4)
	for (i in lines.indices) {
		val line = lines[i]
		val hpw = Math.floor((perpiece / 2f).toDouble()).toInt() - 1

		if (i > 0) {
			for (j in line.indices) {
				var c = ' '
				if (j % 2 == 1) {
					if (line[j - 1] != null) {
						c = if (line[j] != null) '┴' else '┘'
					} else {
						if (j < line.size && line[j] != null) c = '└'
					}
				}
				print(c)

				if (line[j] == null) {
					for (k in 0 until perpiece - 1) {
						print(" ")
					}
				} else {
					for (k in 0 until hpw) {
						print(if (j % 2 == 0) " " else "─")
					}
					print(if (j % 2 == 0) "┌" else "┐")
					for (k in 0 until hpw) {
						print(if (j % 2 == 0) "─" else " ")
					}
				}
			}
			println()
		}

		for (j in line.indices) {
			var f: String? = line[j]
			if (f == null) f = ""
			val gap1 = Math.ceil((perpiece / 2f - f.length / 2f).toDouble()).toInt()
			val gap2 = Math.floor((perpiece / 2f - f.length / 2f).toDouble()).toInt()

			for (k in 0 until gap1) {
				print(" ")
			}
			print(f)
			for (k in 0 until gap2) {
				print(" ")
			}
		}
		println()

		perpiece /= 2
	}
}
